[OCPBUGS-3341] Add RHOSP CCM options reference documentation (#53068)
openshift/openshift-docs: modules/cluster-cloud-controller-config-osp.adoc

| `tenant-id`
| The Identity service project ID. Leave this option unset if you are using Identity service application credentials.

If you are using version 3 of the Identity API, which changed the identifier `tenant` to `project`, the value of `tenant-id` is automatically mapped to the project construct in the API. 

| `tenant-name`
| The Identity service project name. You can ignore this option if `tenant-id` is set.

| `tenant-domain-id`
| The Identity service project domain ID.

| `tenant-domain-name`
| The Identity service project domain name.

| `user-domain-id`
| The Identity service user domain ID.

| `user-domain-name`
| The Identity service user domain name.

| `trust-id`
| The Identity service trust ID. A trust represents the authorization of a user, or trustor, to delegate roles to another user, or trustee. Optionally, a trust authorizes the trustee to impersonate the trustor. You can find available trusts by querying the `/v3/OS-TRUST/trusts` endpoint of the Identity service API.

| `trustee-id`
| The Identity service trustee user ID.

| `trustee-password`
| The Identity service trustee user password.

> | `use-clouds`

Reviewer: @MaysaMacedo We are always testing with below content:

    [Global]
    region      = regionOne
    ca-file     = /etc/kubernetes/static-pod-resources/configmaps/cloud-config/ca-bundle.pem
    use-clouds  = true
    clouds-file = /etc/openstack/secret/clouds.yaml
    cloud       = openstack

If I understood correctly, this configMap is not expected to be changed by the user, so... in which case we can have different configuration than above?

Author: @MaysaMacedo Any thoughts on this?

Reviewer 2: Since we're enforcing the UseClouds to true in the code, I believe there is no need to document the following options: auth-url, os-endpoint-type, username, password, domain related ones, tenant ones and user ones.

@stephenfin @dulek what do you think?

Author: @stephenfin @dulek ^^^

Reviewer 3: I agree with Maysa here.

a| Set this option to `true` to fetch authorization credentials from a `clouds.yaml` file. Options set in this section are prioritized over values read from the `clouds.yaml` file. 

CCM searches for the file in the following places:

. The value of the `clouds-file` option.
. A file path stored in the environment variable `OS_CLIENT_CONFIG_FILE`.
. The directory `pkg/openstack`.
. The directory `~/.config/openstack`.
. The directory `/etc/openstack`.

| `clouds-file`
| The file path of a clouds.yaml file. It is used if the `use-clouds` option is set to `true`.

| `cloud`
| The named cloud in the `clouds.yaml` file that you want to use. It is used if the `use-clouds` option is set to `true`.

> | `application-credential-id`

Reviewer 2: @imatza-rh these application-credentials options are being tested, correct?

Reviewer 4: @MaysaMacedo hey, the `application-credential-id` and `application-credential-secret` options have been tested

| The ID of an application credential to authenticate with. An `application-credential-secret` must be set along with this parameter.

| `application-credential-name`
| The name of an application credential to authenticate with. If `application-credential-id` is not set, the user name and domain must be set.

| `application-credential-secret`
| The secret of an application credential to authenticate with.

| `tls-insecure`
| Whether or not to verify the server's TLS certificate. If set to `true`, the certificate is not verified. By default, the certificate is verified.
|===

[id="ccm-config-networking-options"]

> == Networking options

Reviewer: That is correct. Networking Options were not validated.

| `lb-method`
| The load balancing algorithm used to create the load balancer pool. The value can be `ROUND_ROBIN`, `LEAST_CONNECTIONS`, or `SOURCE_IP`. The default value is `ROUND_ROBIN`.

If you use the `ovn` provider, only the `SOURCE_IP` algorithm is supported.

| `lb-provider`
| Optional. Used to specify the provider of the load balancer, for example, `amphora` or `octavia`. Only the Amphora and Octavia providers are supported.

| `lb-version`
| Optional. The load balancer API version. Only `"v2"` is supported.

| `subnet-id`
| The ID of the Networking servie subnet on which load balancer VIPs are created. 

// This ID is also used to create pool members if `member-subnet-id` is not set.

// | `member-subnet-id`
// | ID of the Neutron network on which to create the members of the load balancer. The load balancer gets another network port on this subnet. Defaults to `subnet-id` if not set.

| `network-id`
| The ID of the Networking service network on which load balancer VIPs are created. Unnecessary if `subnet-id` is set.

// | `manage-security-groups`
// | If the Neutron security groups should be managed separately. Default: false

| `create-monitor`

> | Whether or not to create a health monitor for the service load balancer. A health monitor is required for services that declare `externalTrafficPolicy: Local`. The default value is `false`.

Reviewer 2: perhaps we should mention that when using OVN driver monitor is only supported from OSP17

Reviewer: Right, and there are several restrictions on the LoadBalancer section that were filed as document bugs: 

https://issues.redhat.com/browse/OCPBUGS-2217
https://issues.redhat.com/browse/OCPBUGS-2350
https://issues.redhat.com/browse/OCPBUGS-2789

Reviewer 2: @maxwelldb will those restrictions be included in the release notes or in this docs?

Author: Likely both, as I read them now.

> This option is unsupported if you use the `ovn` provider.

Reviewer 3: This is not true when using OSP17.

Author: ```suggestion
This option is unsupported if you use {rh-openstack} earlier than version 17 and use the `ovn` provider.
```

Reviewer 3: Works for me!

 This option is currently a workaround for the issue https://github.com/kubernetes/ingress-nginx/issues/3996, should be removed or refactored after the Kubernetes [KEP-1860](https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/1860-kube-proxy-IP-node-binding) is implemented.

// | `default-tls-container-ref`
// | Reference to a tls container. This option works with Octavia, when this option is set then the cloud provider will create an Octavia Listener of type TERMINATED_HTTPS for a TLS Terminated loadbalancer.

//  Format for tls container ref: `https://{keymanager_host}/v1/containers/{uuid}`
//  Check `container-store` parameter if you want to disable validation.

// | `container-store`
// | Optional. Used to specify the store of the tls-container-ref, e.g. "barbican" or "external" - other store will cause a warning log. Default value - `barbican` - existence of tls container ref would always be performed. If set to `external` format for tls container ref will not be validated.

| `max-shared-lb`
| The maximum number of Services that can share a load balancer. The default value is `2`.
|===

[id="ccm-config-metadata-options"]

> == Metadata options

Reviewer 2: I don't think these options are tested

Reviewer: True, it was not tested. I think we should remove it from here.

> // Module included in the following assemblies:

Author: I'll attempt to add something that fits this description into #52520. This PR is more about adding the reference module.

> // Module included in the following assemblies:

Author: @stephenfin Thoughts on this? https://github.com/openshift/openshift-docs/pull/52520/commits/e8b486aa585800c8021fce6ee0ae2b6efc7a631a

// | `address-sort-order`
// | This configuration key affects how the provider reports node addresses to Kubernetes node resources. The default order depends on the hard-coded order in which the provider queries addresses and what the cloud returns. A specific order is not guaranteed.

//  To override this behavior, specify a comma-separated list of CIDR addresses. CCM sorts and groups all addresses that match the list in a prioritized manner, wherein the first retrieved item has  a higher priority than the last. Addresses that do not match the list remain in their default order. The default value is `""`.

// This option can be useful if you have multiple or dual-stack interfaces attached to a node that need a user-controlled, deterministic way of sorting addresses.
// |===

[id="ccm-config-lb-options"]
== Load balancer options

CCM supports several load balancer options for deployments that use Octavia. 

[NOTE]
====
Neutron-LBaaS support is deprecated. 
====

// Commented options are untested as of yet.

|===
| Option | Description

| `enabled`
| Whether or not to enable the `LoadBalancer` type of services integration. The default value is `true`.

> | `use-octavia`

Reviewer 2: I would vote to remove this option from the docs or mention that use-octavia is always enforced as true because we don't support anything else than Octavia

> | `use-octavia`
> | Whether or not to use Octavia for the `LoadBalancer` type of service implementation rather than Neutron-LBaaS. The default value is `true`.

Author: ```suggestion
// Removed: always enforced.
//| `use-octavia`
//| Whether or not to use Octavia for the `LoadBalancer` type of service implementation rather than Neutron-LBaaS. The default value is `true`.
```

// Module included in the following assemblies:
//
// * installing/installing_openstack/installing-openstack-cloud-config-reference.adoc

:_content-type: REFERENCE
[id="cluster-cloud-controller-config_{context}"]
= The {rh-openstack} Cloud Controller Manager (CCM) config file

> A {rh-openstack-first} CCM config file, called `cloud.conf` by convention, comprises several options in an INI format.

Reviewer: @stephenfin @MaysaMacedo I think it is not clear, or at least for me, what config file is detailed here. AFAIK, user should modify only the configmap `cloud-provider-config` in the `openshift-config namespace`. Don't you think that should be mention at the top of the doc?

On the other hand, if the intention is to detail the content of the `cloud-conf` configMap in the `openshift-cloud-controller-manager`, all the parameters should be included in the informative section, as the user should never modify that configMap, right?

Reviewer 2: @rlobillo The config detailes here is the user managed config-map present on openshift-config namespace and it's ok if mention that here. I think we should document only the sections that will have an effect if modified by the user, which is what seems to be targeted in the PR

Author: Check out the latest iteration of this PR and LMK.

Reviewer: Hello @maxwelldb I think it's the other way around: The file that the user should modify is the configmap `cloud-provider-config` in the `openshift-config` namespace, and according to @MaysaMacedo and I agree with her, that should be one mentioned on the top of the doc.

Author: @rlobillo See https://github.com/openshift/openshift-docs/pull/53068/files#r1054406909

Reviewer: Hello Max! No, that is not what I meant, sorry, probably I wasn't clear. Let me rephrase starting from the beginning:

1. If the user wants to change something on CCM, it should modify below configMap (`cloud-provider-config` on `openshift-config` namespace):

```
$ oc get configMap/cloud-provider-config --namespace openshift-config -o json | jq -r .data.config 
[Global]
secret-name = openstack-credentials
secret-namespace = kube-system
region = regionOne
ca-file = /etc/kubernetes/static-pod-resources/configmaps/cloud-config/ca-bundle.pem
[LoadBalancer]
lb-provider = amphora
lb-method = SOURCE_IP
create-monitor = True
monitor-delay = 10s
monitor-timeout = 10s
monitor-max-retries = 1
floating-network-id = 03c0b9b2-5994-4611-a7dd-a95df903c2f9
```

2. Then, according to what it is explained on https://github.com/openshift/openshift-docs/pull/52520, the OCP cluster is internally (transparent for the user) converting that new configuration to the configMap called cloud-conf on the `openshift-cloud-controller-manager` namespace:

```
$ oc get configMap/cloud-conf --namespace openshift-cloud-controller-manager -o json | jq -r '.data["cloud.conf"]'
[Global]
region      = regionOne
ca-file     = /etc/kubernetes/static-pod-resources/configmaps/cloud-config/ca-bundle.pem
use-clouds  = true
clouds-file = /etc/openstack/secret/clouds.yaml
cloud       = openstack

[LoadBalancer]
lb-provider         = amphora
lb-method           = SOURCE_IP
create-monitor      = True
monitor-delay       = 10s
monitor-timeout     = 10s
monitor-max-retries = 1
floating-network-id = 03c0b9b2-5994-4611-a7dd-a95df903c2f9
use-octavia         = true
``` 
So, my point is that we should mention at the top of the doc that **if a user want to change something on the cluster, he/she should modify the cloud-provider-config configMap in the openshift-config namespace** and not in the configMap on the openshift-cloud-controller-manager namespace.

Author: :+1: How do you feel about this change, then: https://github.com/openshift/openshift-docs/pull/53068/commits/d73f8fc00b66ea729f698bbc55446c7d1a1fc52b

@rlobillo

Reviewer: that is great, thanks!

| `floating-subnet-id`
| Optional. The external network subnet used to create floating IP addresses for the load balancer VIP. Can be overridden by the service annotation `loadbalancer.openstack.org/floating-subnet-id`.

| `floating-subnet`
| Optional. A name pattern (glob or regular expression if starting with `~`) for the external network subnet used to create floating IP addresses for the load balancer VIP. Can be overridden by the service annotation `loadbalancer.openstack.org/floating-subnet`. If multiple subnets match the pattern, the first one with available IP addresses is used.

| `floating-subnet-tags`
| Optional. Tags for the external network subnet used to create floating IP addresses for the load balancer VIP. Can be overridden by the service annotation `loadbalancer.openstack.org/floating-subnet-tags`. If multiple subnets match these tags, the first one with available IP addresses is used.

If the {rh-openstack} network is configured with sharing disabled, for example, with the `--no-share` flag used during creation, this option is unsupported. Set the network to share to use this option.

| `lb-method`

> | The load balancing algorithm used to create the load balancer pool. The value can be `ROUND_ROBIN`, `LEAST_CONNECTIONS`, or `SOURCE_IP`. The default value is `ROUND_ROBIN`.

Reviewer 5: What about adding the provider here as well?
```suggestion
| The load balancing algorithm used to create the load balancer pool.
For the Amphora provider the value can be `ROUND_ROBIN`, `LEAST_CONNECTIONS`, or `SOURCE_IP`. The default value is `ROUND_ROBIN`.
```

> For the OVN provider, only the `SOURCE_IP` algorithm is supported.

Reviewer 5: The algorithm for the OVN octavia provider is SOURCE_IP_PORT instead of SOURCE_IP.
For reference: https://access.redhat.com/documentation/fr-fr/red_hat_openstack_platform/16.2/html/using_octavia_for_load_balancing-as-a-service/create-other-kinds-lbs_rhosp-lbaas#create-ovn-lb_create-other-kinds-lbs
```suggestion
For the OVN provider, only the `SOURCE_IP_PORT` algorithm is supported.
```

> For the Amphora provider, to use the `LEAST_CONNECTIONS` or `SOURCE_IP` methods, you must configure the `"create-monitor=true` option in the `cloud-provider-config` namespace and `ETP:Local` on the load balancing service.

Reviewer 5: I think we can rephrase this to something like (to make it less mandatory):
```suggestion
For the Amphora provider, when using the `LEAST_CONNECTIONS` or `SOURCE_IP` methods, it's recommended to configure the `"create-monitor=true` option in the `cloud-provider-config` configMap on the `openshift-config` namespace and `ETP:Local` on the load balancer type service in order to get the balancing algorithm enforcement in the client to service endpoint connections.
```

Author: @eurijon Is this the same idea? 

```suggestion
For the Amphora provider, if using the `LEAST_CONNECTIONS` or `SOURCE_IP` methods, configure the `"create-monitor=true` option in the `cloud-provider-config` configMap on the `openshift-config` namespace and `ETP:Local` on the load-balancer type service to allow balancing algorithm enforcement in the client to service endpoint connections.
```

Reviewer 5: @maxwelldb sure it is

| `lb-provider`
| Optional. Used to specify the provider of the load balancer, for example, `amphora` or `octavia`. Only the Amphora and Octavia providers are supported.

| `lb-version`
| Optional. The load balancer API version. Only `"v2"` is supported.

| `subnet-id`

> | The ID of the Networking servie subnet on which load balancer VIPs are created. 

Reviewer 5: Typo
```suggestion
| The ID of the Networking service subnet on which load balancer VIPs are created. 
```

// Module included in the following assemblies:
//
// * installing/installing_openstack/installing-openstack-cloud-config-reference.adoc

:_content-type: REFERENCE
[id="cluster-cloud-controller-config_{context}"]
= The {rh-openstack} Cloud Controller Manager (CCM) config map

> A {rh-openstack-first} CCM config map defines how your cluster interacts with your {rh-openstack} cloud. By default, this configuration is stored under the `cloud.conf` key in the `cloud-conf` config map in the `openshift-cloud-controller-manager` namespace. For example:

Author: ```suggestion
A {rh-openstack-first} CCM config map defines how your cluster interacts with your {rh-openstack} cloud. By default, this configuration is stored under the `cloud.conf` key in the `openshift-cloud-controller-manager` config map in the `cloud-conf` namespace. For example:
```

@rlobillo Is this what you indicated in your comment? 

If so, I need to swap out similar language in #52520. 

I want to be sure, as that other QEed PR and the example config map that I found both have `openshift-cloud-controller-manager` as the namespace. 

@MaysaMacedo @stephenfin

Reviewer: @maxwelldb Nop, this change is not correct. Please discard it.

// Module included in the following assemblies:
//
// * installing/installing_openstack/installing-openstack-cloud-config-reference.adoc

:_content-type: REFERENCE
[id="cluster-cloud-controller-config_{context}"]

> = The {rh-openstack} Cloud Controller Manager (CCM) config map

Reviewer 5: Should we remove RH reference from the CCM name as it's not provided/developed by Red Hat exclusively?
@MaysaMacedo @dulek @stephenfin please correct me if I'm wrong
```suggestion
= The OpenStack Cloud Controller Manager (CCM) config map
```

Reviewer 2: works for me to have only OpenStack

Author: @stephenfin The idea is that I'm required to respect RH branding practices whenever I write about our products, so the PR is littered with "RHOSP" (`{rh-openstack}`). It's muscle memory. :) 

If OpenStack CCM is not subject to RH branding guidelines, I'm free to use "OpenStack" in place of "RHOSP."

My assumption here is that OpenStack CCM is usable outside of Red Hat's OpenStack solution. If that's not the case, let me know.

> A {rh-openstack-first} CCM config map defines how your cluster interacts with your {rh-openstack} cloud. By default, this configuration is stored under the `cloud.conf` key in the `cloud-conf` config map in the `openshift-cloud-controller-manager` namespace. 

Reviewer 5: Same here
```suggestion
A OpenStack CCM config map defines how your cluster interacts with your {rh-openstack} cloud. By default, this configuration is stored under the `cloud.conf` key in the `cloud-conf` config map in the `openshift-cloud-controller-manager` namespace. 
```

[IMPORTANT]
====
The `cloud-conf` config map is generated from the `cloud-provider-config` config map in the `openshift-config` namespace.

To change the settings that are described by the `cloud-conf` config map, modify the `cloud-provider-config` config map. 

> As part of this synchronization, the CCM Operator overrides some options. For more information, see "The {rh-openstack} Cloud Controller Manager".

Reviewer 5: Where can we find the "The {rh-openstack} Cloud Controller Manager" reference? should it be a link?

Author: I have to merge #53068 prior to this PR going live. Link will be included in this document.

Author: This comes from #52520. I can add a link after that's merged.

// The following options are used for {rh-openstack} CCM authentication with the {rh-openstack} Identity service, also known as Keystone. They are similiar to the global options that you can set by using the `openstack` CLI.

// |===
// | Option | Description

// | `ca-file`
// | Optional. The CA certificate bundle file for communication with the {rh-openstack} Identity service. If you use the HTTPS protocol with The Identity service URL, this option is required.

// | `cert-file`
// | Optional. The client certificate path to use for client TLS authentication.

// | `key-file`
// | Optional. The client private key path to use for client TLS authentication.

// | `region`
// | The Identity service region name.

// | `trust-id`
// | The Identity service trust ID. A trust represents the authorization of a user, or trustor, to delegate roles to another user, or trustee. Optionally, a trust authorizes the trustee to impersonate the trustor. You can find available trusts by querying the `/v3/OS-TRUST/trusts` endpoint of the Identity service API.

// | `trustee-id`
// | The Identity service trustee user ID.

// | `trustee-password`
// | The Identity service trustee user password.

> // | `application-credential-id`

Reviewer 2: The options application-credential-id and application-credential-secret were tested by QE and the user is allowed to set it. Could we mention it somewhere?

Reviewer 6: This is kind of by accident. The second we start setting this in `clouds.yaml` there will be a conflict. I'm not actually sure what happens in this situation but I would rather push people to configure stuff via `clouds.yaml` exclusively where possible.

```yaml
clouds:
  foo:
    auth:
      auth_url: <auth-url>
      application_credential_id: "<credential-id>"
      application_credential_secret: "<credential-secret>"
```

wdyt?

Reviewer 2: ah ok, we can leave it out then

| `floating-subnet`
| Optional. A name pattern (glob or regular expression if starting with `~`) for the external network subnet used to create floating IP addresses for the load balancer VIP. Can be overridden by the service annotation `loadbalancer.openstack.org/floating-subnet`. If multiple subnets match the pattern, the first one with available IP addresses is used.

| `floating-subnet-tags`
| Optional. Tags for the external network subnet used to create floating IP addresses for the load balancer VIP. Can be overridden by the service annotation `loadbalancer.openstack.org/floating-subnet-tags`. If multiple subnets match these tags, the first one with available IP addresses is used.

If the {rh-openstack} network is configured with sharing disabled, for example, with the `--no-share` flag used during creation, this option is unsupported. Set the network to share to use this option.

| `lb-method`
| The load balancing algorithm used to create the load balancer pool.
For the Amphora provider the value can be `ROUND_ROBIN`, `LEAST_CONNECTIONS`, or `SOURCE_IP`. The default value is `ROUND_ROBIN`.

For the OVN provider, only the `SOURCE_IP_PORT` algorithm is supported.

> For the Amphora provider, if using the `LEAST_CONNECTIONS` or `SOURCE_IP` methods, configure the `"create-monitor=true` option in the `cloud-provider-config` configMap on the `openshift-config` namespace and `ETP:Local` on the load-balancer type service to allow balancing algorithm enforcement in the client to service endpoint connections.

Reviewer 7: Should there be a closing `"` in here?
```
`"create-monitor=true`
```

Author: Whoops. I'd actually change this to: 

```suggestion
For the Amphora provider, if using the `LEAST_CONNECTIONS` or `SOURCE_IP` methods, configure the `create-monitor` option as `true`  in the `cloud-provider-config` config map on the `openshift-config` namespace and `ETP:Local` on the load-balancer type service to allow balancing algorithm enforcement in the client to service endpoint connections.
```

> For the Amphora provider, if using the `LEAST_CONNECTIONS` or `SOURCE_IP` methods, configure the `"create-monitor=true` option in the `cloud-provider-config` configMap on the `openshift-config` namespace and `ETP:Local` on the load-balancer type service to allow balancing algorithm enforcement in the client to service endpoint connections.

Reviewer 7: ```suggestion
For the Amphora provider, if using the `LEAST_CONNECTIONS` or `SOURCE_IP` methods, configure the `"create-monitor=true` option in the `cloud-provider-config` config map on the `openshift-config` namespace and `ETP:Local` on the load-balancer type service to allow balancing algorithm enforcement in the client to service endpoint connections.
```

| `internal-lb`
| Whether or not to create an internal load balancer without floating IP addresses. The default value is `false`.

// | `cascade-delete`
// | Determines whether or not to perform cascade deletion of load balancers. Default: true.

// | `flavor-id`
// | The id of the loadbalancer flavor to use. Uses octavia default if not set.

// | `availability-zone`
// | The name of the loadbalancer availability zone to use. It is applicable if use-octavia is set to True and requires Octavia API version 2.14 or later (Ussuri release). The Octavia availability zone capabilities will not be used if it is not set. The parameter will be ignored if the Octavia version doesn't support availability zones yet.

| `LoadBalancerClass "ClassName"`
a| This is a config section that comprises a set of options:

 * `floating-network-id`
 * `floating-subnet-id`
 * `floating-subnet`
 * `floating-subnet-tags`
 * `network-id`
 * `subnet-id`

//  * `member-subnet-id`

The behavior of these options is the same as that of the identically named options in the load balancer section of the CCM config file.

> You can set the `ClassName` value by specifying the Service annotation `loadbalancer.openstack.org/class`. 

Reviewer 7: Here and in line 236, should "Service" be lowercase?

Author: ```suggestion
You can set the `ClassName` value by specifying the service annotation `loadbalancer.openstack.org/class`. 
```

// | `ingress-hostname-suffix`
// | The dns suffix to the load balancer IP address when using proxy protocol. Default nip.io

//  This option is currently a workaround for the issue https://github.com/kubernetes/ingress-nginx/issues/3996, should be removed or refactored after the Kubernetes [KEP-1860](https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/1860-kube-proxy-IP-node-binding) is implemented.

// | `default-tls-container-ref`
// | Reference to a tls container. This option works with Octavia, when this option is set then the cloud provider will create an Octavia Listener of type TERMINATED_HTTPS for a TLS Terminated loadbalancer.

//  Format for tls container ref: `https://{keymanager_host}/v1/containers/{uuid}`
//  Check `container-store` parameter if you want to disable validation.

// | `container-store`
// | Optional. Used to specify the store of the tls-container-ref, e.g. "barbican" or "external" - other store will cause a warning log. Default value - `barbican` - existence of tls container ref would always be performed. If set to `external` format for tls container ref will not be validated.

| `max-shared-lb`

> | The maximum number of Services that can share a load balancer. The default value is `2`.

Author: ```suggestion
| The maximum number of services that can share a load balancer. The default value is `2`.
```

|===

// [id="ccm-config-metadata-options"]
// == Metadata options

// |===
// | Option | Description

// | `search-order`
// | This configuration key affects the way that the provider retrieves metadata that relates to the instances in which it runs. The default value of `configDrive,metadataService` results in the provider retrieving metadata that relates to the instance from, if available, the config drive first,and then the metadata service. Alternative values are:
//  * `configDrive`: Only retrieve instance metadata from the configuration drive.
//  * `metadataService`: Only retrieve instance metadata from the metadata service.
//  * `metadataService,configDrive`: Retrieve instance metadata from the metadata service first if available, and then retrieve instance metadata from the configuration drive.
// |===

// ### Multi region support (alpha)

// | environment variable `OS_CCM_REGIONAL` is set to `true` - allow CCM to set ProviderID with region name `${ProviderName}://${REGION}/${instance-id}`. Default: false.

[id="cluster-cloud-controller-config-overrides"]
== Options that the Operator overrides

> The CCM Operator overrides the following options, which you might recognize from configuring {rh-openstack}. Do not configure them yourself. They are included in this document for informational purposes only.

Reviewer 7: Do you think it is worth making the "Do not..." into an admonition? Idk what the consequences are if they miss this.

Author: Your effort is wasted--that's about it, AFAIK. I'm okay leaving as is, as the preceding sentence says as much.

> The CCM Operator overrides the following options, which you might recognize from configuring {rh-openstack}. Do not configure them yourself. They are included in this document for informational purposes only.

Reviewer 2: Looks ok to me

| The Identity service user domain name. 

// This option is not required if you set `domain-id`.

| `tenant-id`
| The Identity service project ID. Leave this option unset if you are using Identity service application credentials.

In version 3 of the Identity API, which changed the identifier `tenant` to `project`, the value of `tenant-id` is automatically mapped to the project construct in the API. 

| `tenant-name`
| The Identity service project name. 

| `tenant-domain-id`
| The Identity service project domain ID.

| `tenant-domain-name`
| The Identity service project domain name.

| `user-domain-id`
| The Identity service user domain ID.

| `user-domain-name`
| The Identity service user domain name.

| `use-clouds`
a| Whether or not to fetch authorization credentials from a `clouds.yaml` file. Options set in this section are prioritized over values read from the `clouds.yaml` file. 

CCM searches for the file in the following places:

. The value of the `clouds-file` option.
. A file path stored in the environment variable `OS_CLIENT_CONFIG_FILE`.
. The directory `pkg/openstack`.
. The directory `~/.config/openstack`.
. The directory `/etc/openstack`.

| `clouds-file`

> | The file path of a clouds.yaml file. It is used if the `use-clouds` option is set to `true`.

Reviewer 7: ```suggestion
| The file path of a `clouds.yaml` file. It is used if the `use-clouds` option is set to `true`.
```

:_content-type: REFERENCE
[id="cluster-cloud-controller-config_{context}"]
= The {rh-openstack} Cloud Controller Manager (CCM) config map

A {rh-openstack-first} CCM config map defines how your cluster interacts with your {rh-openstack} cloud. By default, this configuration is stored under the `cloud.conf` key in the `cloud-conf` config map in the `openshift-cloud-controller-manager` namespace. 

[IMPORTANT]
====
The `cloud-conf` config map is generated from the `cloud-provider-config` config map in the `openshift-config` namespace.

To change the settings that are described by the `cloud-conf` config map, modify the `cloud-provider-config` config map. 

As part of this synchronization, the CCM Operator overrides some options. For more information, see "The {rh-openstack} Cloud Controller Manager".
====

For example:

.An example `cloud-conf` config map
[source,yaml]
----
apiVersion: v1
data:
  cloud.conf: |

>     [Global] <1>

Reviewer 6: ~Is `<1>` an asciidoc thing or a typo?~

Later: Ah, it's an asciidoc thing. I see the callout below.

|===
| Option | Description

> | `ca-file`

Reviewer 6: I actually think every one of these options in the general section is not used by us. They all look like things that would belong in a `cloud.yaml` file and therefore they are things the user should set there (since we're using `clouds.yaml` file), not here. @pierreprinetti @MaysaMacedo wdyt?

Reviewer 2: Good point, Stephen. I would say those are not needed given that we enforce to use cloud.yaml
